### PR TITLE
rc_update: fix on-off-switch with negative threshold values

### LINKED ISSUE #19809: Negative RC threshold not working in v1.13.0
### Describe the bug
I would like to trigger the RTL mode when the switch is at its low value instead of high. In v1.10.1 I could use the parameter [RC_RETURN_TH](https://dev.px4.io/master/en/advanced/parameter_reference.html#RC_RETURN_TH) with a negative value to obtain this behavior. The same parameter in v1.13.0-rc1 has no effect (the RTL mode is not triggered at all when you set a negative value). The workaround is to change the RCX_REV of the corresponding channel but it is not the best solution as this value is cleared when you redo a radio calibration

### To Reproduce
1. Set the Return mode to a channel
2. Change the RC_RETURN_TH to a negative value >= -1
3. The RTL is not triggered anymore

### Expected behavior
Threshold when lower than the value instead of higher 

I guess the issue is the same with all RC mode. Thank you for your help !

### PR DESCRIPTION
**Pull request by MaEtUgR**
### Solved Problem
When reviewing #21763 I wasn't exactly sure why the fix is needed at all see https://github.com/PX4/PX4-Autopilot/pull/21763#issuecomment-1614360730. I tested on hardware and realized it's actually broken and I still don't know why it broke. But I was sure I wanted it to not break the same way again and the function to be more readable. So I added unit testing and fixed it by simplifying the function rather than adding to it.

Fixes #19809
Closes #21763

### Solution
- Improve unit testing of the RCUpdate class I added in #19579
  - adding protected scope again
  - setting parameters from outside the class
- Add unit tests for the return switch
  - Positive threshold cases
  - Corner cases
  - Negative threshold cases that failed
- Fix negative threshold cases

### Changelog Entry
```
Bugfix: Negative RC switch thresholds
```

### Test coverage
Unit tests that I added. I can also test on hardware if desired.

### CODE REVIEW

File: src/modules/rc_update/rc_update.cpp [hunk 1]
```diff
@@ -560,19 +561,16 @@ void RCUpdate::Run()
...
-		if (on_inv ? value < on_th : value > on_th) {
-			return manual_control_switches_s::SWITCH_POS_ON;
-
-		} else {
-			return manual_control_switches_s::SWITCH_POS_OFF;
+		if (threshold < 0.f) {
+			value = -value;
 		}
+
+		return (value > threshold) ? manual_control_switches_s::SWITCH_POS_ON : manual_control_switches_s::SWITCH_POS_OFF;
```
Comment (junwoo091400) [inline src/modules/rc_update/rc_update.cpp:573]: Ohh this is nice refactoring 😋

**Comment by MaEtUgR** [comment]
I made a minimal bugfix port for the 1.14 release: https://github.com/PX4/PX4-Autopilot/pull/21801
Could this pr get reviewed? Then I could merge.

File: src/modules/rc_update/RCUpdateTest.cpp [hunk 2]
```diff
@@ -131,3 +186,60 @@ TEST(RCUpdateTest, ModeSlotButtonAllValues)
...
+TEST_F(RCUpdateTest, ReturnSwitchNegativeThresholds)
+{
+	checkReturnSwitch(1.f, -0.5f, 3); // Above threshold -> SWITCH_POS_OFF
+	checkReturnSwitch(0.f, -0.5f, 3); // On threshold -> SWITCH_POS_OFF
+	checkReturnSwitch(-.001f, -0.5f, 1); // Slightly below threshold -> SWITCH_POS_ON
+	checkReturnSwitch(-1.f, -0.5f, 1); // Below threshold -> SWITCH_POS_ON
+
+	checkReturnSwitch(1.f, -0.75f, 3); // Above threshold -> SWITCH_POS_OFF
+	checkReturnSwitch(.5f, -0.75f, 3); // On threshold -> SWITCH_POS_OFF
+	checkReturnSwitch(-.001f, -0.75f, 1); // Slightly below threshold -> SWITCH_POS_ON
```
Comment (junwoo091400) [inline src/modules/rc_update/RCUpdateTest.cpp:236]:
```suggestion
	checkReturnSwitch(.499f, -0.75f, 1); // Slightly below threshold -> SWITCH_POS_ON
```

I think in general it's quite confusing having two different scales (-1 ... +1 for channel values & 0 ... 1 for threshold), but anyhow this value needs to be changed 🤔

File: src/modules/rc_update/rc_update.cpp [hunk 3]
```diff
@@ -123,8 +123,10 @@ bool RCUpdate::init()
 	return true;
 }
 
-void RCUpdate::parameters_updated()
```
Comment (junwoo091400) [inline src/modules/rc_update/rc_update.cpp:126]: Wow so until now when parameters were updated, this function was actually never called? (Since it's not overriding `updateParams`)

File: src/modules/rc_update/RCUpdateTest.cpp [hunk 4]
```diff
@@ -38,15 +38,120 @@
...
 using namespace rc_update;
 
-TEST(RCUpdateTest, ModeSlotUnassigned)
+class TestRCUpdate : public RCUpdate
+{
+public:
+	void UpdateManualSwitches(const hrt_abstime &timestamp_sample) { RCUpdate::UpdateManualSwitches(timestamp_sample); }
+	void updateParams() { RCUpdate::updateParams(); }
+	void setChannel(size_t index, float channel_value) { _rc.channels[index] = channel_value; }
+};
```
Comment (junwoo091400) [inline src/modules/rc_update/RCUpdateTest.cpp:47]: Is this class really needed? The only function it adds is `setChannel` right?

**Review by junwoo091400 — CHANGES_REQUESTED** [review]
Seems nice, thanks for this follow-up PR! Let's get this in `main`, then into `1.14` 🔥

File: src/modules/rc_update/RCUpdateTest.cpp (hunk 2 above)
Comment (MaEtUgR) [reply src/modules/rc_update/RCUpdateTest.cpp:236]: Right you are! Thanks for looking into the details 🙏

File: src/modules/rc_update/rc_update.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/rc_update/rc_update.cpp:126]: It was manually called before otherwise stuff would not have worked properly. I removed the separate call since it's now going through the `updateParams()` mechanism. See: https://github.com/PX4/PX4-Autopilot/commit/76289d92f24d04bed89a74b9ca1208b586f10bc9#diff-bc65e8936281291b7e9801b185ee1740a6f2d4207e3b539c059469daa6cdfcc9L391

File: src/modules/rc_update/RCUpdateTest.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/rc_update/RCUpdateTest.cpp:47]: Well in https://github.com/PX4/PX4-Autopilot/commit/4bd2d4cf3576b5f87c761a485420a814f3084cec#diff-5c02ad0b2293fe7ee164aea9fe5387fba87e6305fa543acc9f08a6a5cc09716cL93 I removed the private access specifier of member variables to be able to mutate them from the testing. In this pr I'm putting back the protected access specifier and this class is used as kind of a test "friend" explicitly exposing what's needed for testing. The test and class should be restructured to cover all testing without mutating internals. This helper class is an in-between incremental step showing what's missing and can be removed again when doing these changes.

Comment (junwoo091400) [reply src/modules/rc_update/RCUpdateTest.cpp:47]: Oh gotcha, didn't notice that those were `protected` functions!

File: src/modules/rc_update/rc_update.cpp (hunk 3 above)
Comment (junwoo091400) [reply src/modules/rc_update/rc_update.cpp:126]: I'm often confused about this `updateParams` (I guess the correct way to do it is by overriding this, as handled already in ModuleBase) or `parameters_updated`, or `update_parameters` scattered around the codebase 😆

**Review by junwoo091400 — APPROVED** [review]
